### ISSUE TITLE
Pi OS changes resulted in "System initializing. Not receiving data."

### ISSUE THREAD

**Issue by daved44** [jgyates/genmon#982]
My Genmon installation has been running fine for a few weeks and yesterday I made some changes and updates, after which the above message appears. The Genmon status page displays a value only in the CPU Temp gauge; the remaining indicators are zero. Most updates were done from my desktop, SSH to Genmon. I have not removed or disturbed the Pi/HAT unit in the generator.  Most of what I've done I first tried on a second Pi 4 I call PlayPi located at my desktop running the same Bookworm as the Genmon Pi. BTW, I always do a reboot after making a change or shutdown and remove power before swapping SD cards.

I booted the Genmon Pi from an earlier SD card with an smaller Bookworm OS image. It runs the Genmon code fine, so there is no hardware issue.

What I've done:
1. I began using rpi-clone to do  regular resyncs several days ago after installing the version from https://github.com/framps/rpi-clone that fixed a problem where clones or resyncs of Bookworm images made with rpi-clone would not boot because of a boot configuration change in this OS version. All the resyncs booted and checked out OK, or so I thought.  I have found an issue but can't locate it now stating that an rpi-sync may miss changed files if the date wasn't changed. rpi-clone apparently does not do a character-by-character comparison of files for change. But I do not believe that rpi-sync was causing any problems, especially since I continued using the primary SD card in the Genmon Pi. 
2.  I got RealVNC working to my desktop PC and a Windows tablet on both PlayPi and Genmon. RealVNC does not yet support the Wayland GUI version that Bookworm defaults to, so I had to switch to X11. That all works OK and allows me to get to the Genmon OS via GUI. 
- I switched back to Wayland on the Genmon OS and rebooted. This did not help.
3.  A bunch of updates, including firmware, installed on the Genmon OS. 
4.  I renamed the host on the Genmon Pi to Genmon.  
- I have since changed it back to what it was and it did hot help.
5.  I had to enable BT on Genmon several days ago to fix an issue documented on this site. Thinking that this could be related, I ran the several commands again, beginning with  sudo python3 serialconfig.py -e -b. That didn't help.
6.  The backup card does almost the same thing, meaning that I must have done a resync after the problem appeared on Genmon.  Almost, but not exactly. On the Genmon SD card, there are no longer any entries in the Genmon configuration, such as email ID, Add-ons I've installed, etc. The backup does have my configuration as it should be, but there is no connection to the generator. 

I would like to fix the failing image, partly because I made a lot of changes to it that I would like to keep and partly because this is a Linux learning project for me; I don't want to declare defeat on it!

Environment
Generator Model: 20kW Evolution 1 Generac
Genmon Logs: Submitted via the About page
Genmon Version: V1.19.00
Pi 4 running full GUI version of Bookworm
PintSize.Me HAT

**Comment by jgyates** [comment]
I got you logs. The underlying issues is that your serial comms are not working. If you check on the Monitor page you will see 'Packet Count" under Comm Stats. You logs show this:

    Packet Count : M: 113, S: 0

This means the pi attempted to send 113 packets (M for Master) and got none back zero (S for slave).  Enabling BT uses one of the serial ports so enabling the serial port needs to factor in the system is using BT.

I am guessing that the issues is related to enabling BT. If you run the serialconfig.py setup first without BT, then enabled it you must restore the settings before running it again to enable the serial port with BT. Try this:

     cd genmon
     cd OtherApps
      python3 serialconfig.py -r

This will restore the original boot settings for the serial port back to the default OS settings. Now enable the serial port with BT enabled as well.

      python3 serialconfig.py -e -b

then reboot and see if that fixes the issue. If it does not then send me your logs again once the system has rebooted. Also if it does not fix the issue then show the output of this command


      ls -la /dev/ser*

**Comment by jgyates** [comment]
The instructions for restoring from a genmon backup are on this page:

https://github.com/jgyates/genmon/wiki/Appendix-M-Tips-and-Tricks-useful-for-Genmon

Genmon needs a file system, a serial port or serial over TCP connection and a network connection. Other than that is does not matter if your are running a GUI from genmon's perspective. A GUI can shorten the life of a SD card, compared to just running linux and no GUI. In general writes to an SD card shorten the life of a SD card, but the expected life span is still around 10 years.

I don't know anything about rpi-clone so I don't have anything to add there.

I am going to close this issue but post any follow up questions to this thread.

**Comment by daved44** [comment]
I don't have an SD card problem or a problem in the Genmon Pi. Running on new SD cards, I tried your above suggested BT changes  with the image running in the inside PlayPi, rebooted, and it failed with the same  "BusyBox V1.35 (Debian 1:1.35.0-4+B3) built-in shell (ash)" message. So serialconfig.py is doing some harm.  The PlayPi is the same Pi 4, but with more memory.

**Comment by jgyates** [comment]
Has this system has the OS upgraded or just installed software updates? What changes and updates were made?

If you have upgraded your debian version that could be an issue. In my experience upgrading from one debian release to another can be problematic, especially if you use a graphical UI.

You can restore your settings with 'python3 serialconfig.py -r' and then just use raspi-config to make the changes with the built in tool to enable the serial port and bluetooth and see if that works. Just make a copy of your original /boot/config.txt and /boot/cmdline.txt (these are the only two files that serailconfig.py modifies).

**Comment by daved44** [comment]
With a bit of snooping around after loading the failed OS SD card into PlayPi running the working backup version, I discovered what may be the reason for the failed boot. I had to work through what apparently is the same problem to make rpi-clone work with Bookworm when a copy made with rpi-clone would not boot, in the same manner.  See https://github.com/billw2/rpi-clone. issues #168 and #169.  A quote from 168: "The official rpios bookworm release has changed the mount point /boot to /boot/firmware." The fix to rpi-clone by framps is documented in #169, and I replaced the original rpi-clone with his version from https://github.com/framps/rpi-clone; it has worked properly for me since. 

In Bookworm, /boot has links to /boot/config.txt, /boot/cmdline.txt, and the /boot/overlays folder, which reside in the /boot/firmware folder .  Those links are severed in the failed copy I'm looking at:
![image](https://github.com/jgyates/genmon/assets/148695933/5b6677ad-2ba2-4106-a0e2-2e7a6e2ebfec)

To answer your questions above, I built both the working and failing Bookworm OS images from scratch by imaging new SD cards with the Windows Raspberry Pi Imager. 

The working image was loaded from the recommended topmost 64-bit GUI 1.1GB version and is essentially unaltered with nothing much added except Genmon. I haven't run your BT fix code; BT remains unavailable on that image, and Genmon works as it should.

The failed image is a larger Bookworm image, accessed from Raspberry Pi OS (other)>>Raspberry Pi OS Full (64-bit), 1.7GB.  I see no essential differences between the two; the larger image simply has more applications and toys. 

But I added and changed many things on the failing image:
- Fixed BT with your fix. It booted correctly with that fix;
- Added the large gnome package, for lots of additional apps, mostly a real task manager (I did not know then or now whether gnome made massive changes to my OS);
- added rpi-clone and the Bookworm framps fix;
- activated RealVNC, for which I had to change the Wayland backend to X11, since RealVNC does not yet support Wayland but this backend is standard with Bookworm;
- perhaps more.

I've had no problems backing up the working image, but I always did it with the GUI SD Card Copy utility from the Pi I have inside.

I suspect that you want to fix the Bookworm issue and I am happy to provide as much looking/testing as you like. But in the meantime I plan to use the working version, updating as needed (mostly RealVNC and rpi-clone) and fully testing after each change. Backing up after each change, I should be able to avoid big problems.

**Comment by jgyates** [comment]
Hi @daved44 

I am guessing the issue here is the rpi-clone breaks the links of the files in /boot rather than serialconfig.py. One of the reasons I say this is that it looks like the overlays file also has it's link broken. serialconfig.py does not touch that file.  rpi-clone uses rsync, which checks for modified files. Once serialconfig.py modifies the files then rpi-clone (via rsync) will attempt to do it thing which is were, it appears, the links get broken. 

In my testing on bookworm serialconfig.py works as expected. I also took a quick look at the source for the two rpi-clone repositories and they are using rsync with some partition manipulation. serialconfig just edits two text files in the boot folder, which works fine even with the new bookworm link.

If you are referring to another bookworm issues let me know. I do want to fix any bookworm issues with genmon, but at the moment it is not obvious to me that serialconfig.py has an issues with bookworm. Let me know if I am missing something.

**Comment by daved44** [comment]
That makes sense since, as I pointed out above, I had used your serialconfig.py successfully on my larger image, and I did that before I put on the rpi-clone fix. 

I will remove the rpi-clone install from my list of ToDo's  on the working image, at least for the time being.  One of the problems with that project is the author hasn't been seen for several years and other volunteers have been fixing problems, with lots of forks that may go nowhere. 

With that in mind, have you any suggestions for micro-SD card copying with perhaps resync capability, done in terminal or GUI mode?

One of the things I need to do is to change the hostname on the working image. I read somewhere that I also must change this to Genmon, but cannot find that item now. Where do I change and is this idiot proof, or should I leave hostname as is?

**Comment by daved44** [comment]
I got the X11 headless GUI working, not difficult to do. That means that I can hang a backup SD card on the Genmon Pi and use the SD Card Copy program, doing a complete image each time, since I'm in no hurry to pull it out and since the image this utility produces can be trusted to be correct, plus I can examine the contents online. I can leave it there and backup when necessary, swapping it out for a second backup card from time to time.  rpi-clone will remain a programma-non-grata. 

That also cements our belief that rpi-clone caused the problem, not anything in Genmon or serialconfig.py.

**Comment by daved44** [comment]
I changed the hostname with raspi-config per your instructions, rebooted, and genmon did not connect, same Monitor status of 100% timeout errors. I changed it back to the original hostname, rebooted. It still will not connect. Remember, this is Bookworm; what used to work may not work now...

**Comment by jgyates** [comment]
send your logs, that may have some info

**Comment by jgyates** [comment]
I just tested this on pi 4 running bookworm and it is working for me. Serial communications does not have any dependancies on network name or the network in general. Has your serial device names changed?

**Comment by daved44** [comment]
Serial device names? Where do I find that and how do I know if changed? Perhaps I should again run serialconfig.py? Remember, this image does not have BT enabled, as I have never run serialconfig.py to enable it, and don't plan to unless necessary now. 

The only thing done to this image after installing genmon is activation of RealVNC, which required changing from Wayland to X11 backends (raspi-config>>Advanced Options>>Wayland. It works fine with that addition.

**Comment by jgyates** [comment]
Your logs don't show any issues with the serial device. 

The serial device names can be show with this command:

      ls -la /dev/ser*

the output will show /dev/serial0 or /dev/serial1 and that must match the "Port for Serial Communication" on the settings page.

Were any other changes made other than hostname?

**Comment by daved44** [comment]
Port for Serial Communications is /dev/serial0. Looking at /dev, serial0 is a link to target file ttyAMA0. 
I made no other changes than what I noted above, activating RealVNC with the required change from Wayland to X11 backend and getting RealVNC to start up at boot.

I really have only a couple things left to do from a working image: Restore the backed-up configuration tar.gz file, .conf only, PW change, and this hostname change. Should I go ahead and restore the working backup and go forward with the other things or wait a bit while you research farther?

Have you tried changing from Wayland to X11? and enabling VNC? It looks like there is quite a bit there.

**Comment by jgyates** [comment]
The logs do not show any serial device errors. Try power cycling the pi (remove power and reapply power), if that does not work then we should fall back to a loopback test. The pi things it is sending it is just not receiving, according to the logs.

There are some errors in some of the add on logs that look like this:

      Error in IsRunning: PID still exists but it's a zombie 

This is an error that each program (genmon, and add ons) check to see if that particular module is already loaded. If it is then it does not load. This should not happen if restarts occurd as they should (shutdown and start with the script).

If a power cycle does not fix the issue then send the logs again after the reboot and then run a loopback test.

I don't run a graphical UI on any of the pi's I have setup. I have one that is connected to my generator and I have 3 dedicated to testing genmon. I have 5 others connected to various other things. They all run headless with ssh access only. The logs do not indicate any issue with the device name, or another device trying to access the serial port (if X11 was loading something that would access the serial port that would cause another error besides timeouts). Because of this the loopback would provide more useful info based on the data in the logs (only timeout errors), assuming a reboot does not fix the issue.

**Comment by daved44** [comment]
I restored the primary SD card and booted it; Genmon did not work. Instead of wasting a lot more of your time (and mine), it will be much quicker to start from scratch by imaging a new OS, the 64-bit Bookworm GUI smaller version, which is the one recommended by the Raspberry Pi Imager, while getting the hostname, user name, etc. correct in the first place. Then I will put Genmon on it, restore the configuration, back it up, then move forward a bit at a time, documenting each tiny change, and backing up at each step, alternating to a couple backup SD cards.

The problem must be something I did or an improper backup. This won't take too long, much less if I have to start finding my way around doing a loopback test along with all of this additional diagnosis, and I'll know where I am at each step.  I should get to a working system set up with RealVNC by EOD.
